Add branch and actor filters to `run list`

bak1an opened the pull request:
Both https://docs.github.com/en/rest/reference/actions#list-workflow-runs and https://docs.github.com/en/rest/reference/actions#list-workflow-runs-for-a-repository accept those query parameters and seems that it would be handy to have it in cli as well.

<hr/>

upd: Just noticed that there is a related issue #3686, have not seen it initially when opening the PR. Will pause adding tests until getting some initial feedback on the approach taken.

bak1an committed:
Add branch and actor filters to `run list`

cliAutomation commented:
Hi! Thanks for the pull request. Please ensure that this change is linked to an issue by mentioning an issue number in the description of the pull request. If this pull request would close the issue, please put the word 'Fixes' before the issue number somewhere in the pull request body. If this is a tiny change like fixing a typo, feel free to ignore this message.

sarcasticadmin commented:
Anything that still needs to be done before we can get this merged? Id love to utilize this :)

bak1an commented:
@mislav  Can this be moved forward somehow?

samcoe commented on pkg/cmd/run/shared/shared.go:
```diff
@@ -213,6 +231,7 @@ func getRuns(client *api.Client, repo ghrepo.Interface, path string, limit int)
 		query := parsed.Query()
 		query.Set("per_page", fmt.Sprintf("%d", perPage))
 		query.Set("page", fmt.Sprintf("%d", page))
+		opts.updateQuery(&query)
```
I think I would rather see the `getRuns` function be responsible for constructing the query instead of offloading the work to the `FilterOptions`. 
```golang
if opts != nil {
	if opts.Branch != "" {
 		query.Set("branch", opts.Branch)
 	}
 	if opts.Actor != "" {
 		query.Set("actor", opts.Actor)
 	}
}
```
Additionally adding the `nil` check here would allow us to remove the `EmptyFilters()` function and just pass in `nil` when we don't want to filter by branch or actor.

bak1an replied:
Sounds good to me, I'll update the PR

samcoe reviewed (commented):
@bak1an Thanks for the contribution and sorry for the delay getting this reviewed. Overall really like the approach here I just left one comment about reducing the responsibilities of `FilterOptions`. Are you still open to adding tests for this? Seems like you were waiting for direction approval before adding those.

bak1an commented:
@samcoe Thanks for a review! I am still around and can update the PR and will try adding tests for this.

bak1an committed:
Simplify what FilterOptions can do

bak1an committed:
Check not only limit in TestNewCmdList

bak1an committed:
Verify that branch/actor params are parsed properly

bak1an committed:
Verify that API requests have proper query parameters

bak1an commented:
@samcoe Hi. The PR is now rebased onto fresh trunk, `FilterOptions` does not do anything smart and added some tests.

Added checks in `TestNewCmdList` that corresponding command line parameters are properly stored into `ListOptions`.

Also checking in `TestListRun` that stubbed API requests do have corresponding query parameters when they are set in given `ListOptions`. I added ad hoc `httpmock.Matcher` right inside the test to do that. Was not sure if it will be useful for anything else and decided not to touch `httpmock` itself to keep this PR more about runs filtering.

samcoe committed:
Change flag name from actor to user

samcoe reviewed (approved):
@bak1an This is great! Thanks for being thorough with the tests. I like the new query string matcher a lot and I can see it being used in the future. When that happens we can extract it into `httpmock`. I pushed one small commit to just change the name of the flag from `actor` to `user`. Thanks for your contribution.

bak1an commented:
Thanks @samcoe 🎉
